zookeeper/dataset.py

```diff
@@ -0,0 +1,138 @@
+    train_split: str
+    validation_split: Optional[str] = None
+
+    def validate_configuration(self):
+        super().validate_configuration()
+
+        # Check that the name corresponds to a valid TensorFlow dataset.
+        builder_names = tfds.list_builders()
+        if self.name not in builder_names:
+            raise ValueError(
+                f"'{self.__component_name__}.name' has invalid value '{self.name}'. "
+                "Valid values:\n    " + ",\n    ".join(builder_names)
+            )
+
+        # Check that the `train_split` is valid.
+        if self.train_split is None or any(
+            s not in self.splits for s in base_splits(self.train_split)
+        ):
+            raise ValueError(
+                f"'{self.__component_name__}.train_split' has invalid value "
+                f"'{self.train_split}'. Valid values:\n    "
+                + ",\n    ".join(self.splits.keys())
+            )
+
+        # Check that `validation_split` is valid (`None` is allowed).
+        if self.validation_split is not None and any(
+            s not in self.splits for s in base_splits(self.train_split)
+        ):
+            raise ValueError(
+                f"'{self.__component_name__}.train_split' has invalid value "
+                f"'{self.train_split}'. Valid values:\n    "
+                + ",\n    ".join([None] + self.splits.keys())
+            )
+
+    @property
+    def name_version(self):
+        return f"{self.name}:{self.version}" if self.version else self.name
+
+    @property
+    def info(self):
+        if not hasattr(self, "_info"):
+            self._info = tfds.builder(self.name, data_dir=self.data_dir).info
+        return self._info
+
+    @property
+    def splits(self):
+        return self.info.splits
+
+    @property
+    def num_classes(self) -> int:
+        try:
+            features = self.info.features
+            if "label" in features:
+                return features["label"].num_classes
+            if "labels" in features:
+                return features["labels"].feature.num_classes
+            if "objects" in features and "label" in features["objects"]:
+                return features["objects"]["label"].num_classes
+        except Exception:
+            pass
+        raise ValueError("Unable to determine the number of classes.")
+
+    def num_examples(self, split) -> int:
+        """Compute the number of examples in a given split."""
+
+        return sum(self.splits[s].num_examples for s in base_splits(split))
+
+    def load(self, split, decoders, shuffle) -> tf.data.Dataset:
+        """Return a `tf.data.Dataset` object representing the requested split."""
+
+        data = tfds.load(
+            name=self.name_version,
+            split=split,
+            data_dir=self.data_dir,
+            decoders=decoders,
+            as_dataset_kwargs={"shuffle_files": shuffle},
+        )
+        return data.cache() if self.cache_in_memory else data
```

Comment:
What would the tradeoff be when caching in userland instead? Or should this be an argument to the method?

Reply:
In effect this is an argument to the component, because `cache_in_memory` is a parameter. You're right though that it would be more flexible to remove it and just return the dataset directly for the user to manipulate.